packages/tre/src/plugins/cache/gateway.ts

```diff
@@ -130,137 +142,186 @@ function getMatchResponse(
+    const baseURL = await this.#ready;
     // Since multiple parses can be in-flight at once, an identifier is needed
     const id = `/${crypto.randomBytes(16).toString("hex")}`;
-    this.responses.set(id, response);
-    const address = this.server.address()! as AddressInfo;
+    this.#responses.set(id, response);
     try {
-      const parsedResponse = await fetch(
-        `http://localhost:${address.port}${id}`
-      );
-      const body = await parsedResponse.arrayBuffer();
-      return {
-        headers: parsedResponse.headers,
-        status: parsedResponse.status,
-        body: new Uint8Array(body),
-      };
+      return await fetch(new URL(id, baseURL));
     } finally {
-      this.responses.delete(id);
+      this.#responses.delete(id);
     }
   }
 }
 
+class SizingStream extends TransformStream<Uint8Array, Uint8Array> {
+  readonly size: Promise<number>;
+
+  constructor() {
+    const sizePromise = new DeferredPromise<number>();
+    let size = 0;
+    super({
+      transform(chunk, controller) {
+        size += chunk.byteLength;
+        controller.enqueue(chunk);
+      },
+      flush() {
+        sizePromise.resolve(size);
+      },
+    });
+    this.size = sizePromise;
+  }
+}
+
 export class CacheGateway {
+  private readonly storage: KeyValueStorage<CacheMetadata>;
+
   constructor(
     private readonly log: Log,
-    private readonly storage: Storage,
+    legacyStorage: Storage,
     private readonly clock: Clock
-  ) {}
+  ) {
+    const storage = legacyStorage.getNewStorage();
+    this.storage = new KeyValueStorage(storage, clock);
+  }
 
   async match(request: Request, cacheKey?: string): Promise<Response> {
     // Never cache Workers Sites requests, so we always return on-disk files
     if (isSitesRequest(request)) throw new CacheMiss();
-
     cacheKey ??= request.url;
-    const cached = await this.storage.get<CacheMetadata>(cacheKey);
+
+    let resHeaders: Headers | undefined;
+    let resRanges: InclusiveRange[] | undefined;
+
+    const cached = await this.storage.get(cacheKey, ({ size, headers }) => {
+      resHeaders = new Headers(headers);
+      const contentType = resHeaders.get("Content-Type");
+
+      // Need size from metadata to parse `Range` header
+      const rangeHeader = request.headers.get("Range");
+      if (rangeHeader !== null) {
+        resRanges = _parseRanges(rangeHeader, size);
+        if (resRanges === undefined) throw new RangeNotSatisfiable(size);
+      }
+
+      return {
+        ranges: resRanges,
+        contentLength: size,
+        contentType: contentType ?? undefined,
+      };
+    });
     if (cached?.metadata === undefined) throw new CacheMiss();
 
-    const response = new CacheResponse(
-      cached.metadata,
-      cached.value
-    ).toResponse();
-    response.headers.set("CF-Cache-Status", "HIT");
+    // Avoid re-constructing headers if we already extracted multipart options
+    resHeaders ??= new Headers(cached.metadata.headers);
```

Comment:
Is there ever a case where `resHeaders` won't have been set on line 247? It seems like the options function is caled unconditionally?